lib/index.js

```diff
@@ -0,0 +1,281 @@
+ */
+async function encryptForDid(did, message, expiration) {
+  const didOptions = {did, location: 'ledger', mode: 'test'};
+  const cipherKeys = await cipherKeysForDid(didOptions);
+
+  const cipher = encryptMessage(message, cipherKeys);
+
+  return {
+    expiration: parseExpiration(expiration), // convert to seconds
+    keyId: cipherKeys.keyId,
+    cipher: bs58.encode(cipher)
+  };
+}
+
+/**
+ * @param message {string} Plaintext message
+ * @param keys {KeyPair} curve25519 key pair
+ *
+ * @returns {Buffer} Message encrypted via libsodium's crypto_box_seal
+ */
+function encryptMessage(message, keys) {
+  message = new Buffer(message);
+  const cipher = new Buffer(message.length + sodium.crypto_box_SEALBYTES);
+  sodium.crypto_box_seal(cipher, message, keys.publicKey);
+
+  return cipher;
+}
+
+/**
+ * @param message {string}
+ * @param storeUrl {string}
+ * @returns {Promise<string>}
+ */
+function saveMessage(message, storeUrl) {
+  const url = storeUrl.endsWith('/whisper')
+    ? storeUrl
+    : storeUrl + '/whisper';
+
+  return axios.post(url, message, {responseType: 'text'})
+    .then(res => res.data);
+}
+
+/**
+ * @param url {string} URL of stored message, or DID of recipient
+ * @param [cipher] {string} Encrypted message, base58-encoded
+ *
+ * @returns {Promise<string>} Resolves with decrypted message
+ */
+function decrypt(url, cipher) {
+  if(cipher) {
+    const did = url;
+    return decryptForDid(did, cipher);
+  }
+  // Common case, fetch cipher from a did-whisper-server instance
+  return decryptStored(url);
+}
+
+/**
+ * @param url {string} URL of stored message
+ *
+ * @returns {Promise<string>} Resolves with decrypted message
+ */
+async function decryptStored(url) {
+  const storedMessage = await fetchMessage(url);
+
+  const {cipher, keyId} = storedMessage;
+  const did = didFromKeyId(keyId);
+
+  return decryptForDid(did, cipher, keyId);
+}
+
+/**
+ * @param did {string} DID of recipient
+ * @param cipher {string} Encrypted cipher text, base58-encoded
+ * @param [keyId] {string}
+ *
+ * @returns {Promise<string>} Decrypted plaintext message (using authentication
+ *   private and public keys in the DID Document)
+ */
+async function decryptForDid(did, cipher, keyId) {
+  const didOptions = {did, location: 'local', mode: 'test'};
+  const recipientKeys = await cipherKeysForDid(didOptions);
+
+  if(!recipientKeys.secretKey || !recipientKeys.publicKey) {
+    throw new Error('Could not find private decryption key in DID Document');
+  }
+
+  cipher = bs58.decode(cipher);
+
+  return decryptMessage(cipher, recipientKeys);
+}
+
+function decryptMessage(cipher, recipientKeys) {
+  const message = new Buffer(cipher.length - sodium.crypto_box_SEALBYTES);
+  sodium.crypto_box_seal_open(message, cipher, recipientKeys.publicKey,
+    recipientKeys.secretKey);
+
+  return message.toString();
+}
+
+/**
+ * @param url {string}
+ *
+ * @throws {Error} Encountered while fetching or parsing stored message
```

Comment:
I think I meant it as a reminder that any sort of network operation might result in an error. My intention was that those should be handled by the caller code, rather than here, but am open to suggestions.